lib/service/service.go

```diff
@@ -3203,8 +3203,25 @@ func (process *TeleportProcess) initDiagnosticService() error {
 	logger.InfoContext(process.ExitContext(), "Starting diagnostic service.", "listen_address", process.Config.DiagnosticAddr.Addr)
 
 	process.RegisterFunc("diagnostic.service", func() error {
-		err := server.Serve(listener)
-		if err != nil && err != http.ErrServerClosed {
+		muxListener, err := multiplexer.New(multiplexer.Config{
+			Context:                        process.GracefulExitContext(),
+			Listener:                       listener,
+			PROXYProtocolMode:              multiplexer.PROXYProtocolUnspecified,
+			SuppressUnexpectedPROXYWarning: true,
+			ID:                             teleport.Component(teleport.ComponentDiagnostic),
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		listenerHTTP := muxListener.HTTP()
+		go func() {
+			if err := muxListener.Serve(); err != nil && !utils.IsOKNetworkError(err) {
+				muxListener.Entry.WithError(err).Error("Mux encountered err serving")
+			}
+		}()
+
+		err = server.Serve(listenerHTTP)
```

Comment:
With this change we close `listenerHTTP` rather than `listener` when we close `server`; I suspect that it's not a problem in practice (and it might be slightly more correct, even, considering that we want to manage the lifetime of the listener sockets in `TeleportProcess` rather than let random goroutines close them) but it might be worth noting in a comment in `diagnostic.shutdown`.

Reply:
We now close multiplexer in `diagnostic.shutdown`, which closes `listener`. So I guess we can do without the comment, since we've got virtually the same situation?